### ISSUE TITLE
AssertionError thrown when generating

### ISSUE THREAD

**Issue by johnscancella** [Yelp/git-code-debt#62]
Every time I try to generate the code debt it throws an assertion error
```
$ git-code-debt-create-tables database.db
$ git-code-debt-generate git@github.com:LibraryOfCongress/bagit-java.git database.db
Cloning into '/var/folders/bl/xmfqp4yx4z1607wydb19x0dc0000gn/T/tmpcYxf4wtemp-repo'...
remote: Counting objects: 26904, done.
remote: Total 26904 (delta 0), reused 0 (delta 0), pack-reused 26904
Receiving objects: 100% (26904/26904), 145.16 MiB | 10.22 MiB/s, done.
Resolving deltas: 100% (11730/11730), done.
Checking connectivity... done.
Traceback (most recent call last):
  File "/usr/local/bin/git-code-debt-generate", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/git_code_debt/generate.py", line 180, in main
    args.skip_default_metrics,
  File "/usr/local/lib/python2.7/site-packages/git_code_debt/generate.py", line 97, in load_data
    commits, pool.imap(_get_metrics_inner, mp_args),
  File "/usr/local/Cellar/python/2.7.10_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/pool.py", line 668, in next
    raise value
AssertionError
```

**Comment by asottile** [comment]
Interesting, this is probably due to the special file handling.  Could you also provide what version of git you are using in case I can't reproduce when I get to a computer?

**Comment by johnscancella** [comment]
sure, I am using git version 2.9.0

**Comment by dannyroberts** [comment]
I'm seeing the same traceback, I'm on git 2.9.3

**Comment by asottile** [comment]
- Succeeds for me using git 1.9.1
- Fails for me on git 2.10.0-rc1 (with the same trace), let's see what changed!

**Comment by asottile** [comment]
With this patch I get a better stacktrace:

```diff
$ git diff
diff --git a/git_code_debt/generate.py b/git_code_debt/generate.py
index 56c5bc4..5b1adb4 100644
--- a/git_code_debt/generate.py
+++ b/git_code_debt/generate.py
@@ -94,7 +94,7 @@ def load_data(
             )
             pool = multiprocessing.pool.Pool(15)
             for commit, metrics in six.moves.zip(
-                    commits, pool.imap(_get_metrics_inner, mp_args),
+                    commits, map(_get_metrics_inner, mp_args),
             ):
                 increment_metric_values(metric_values, metrics)
                 insert_metric_values(
```

(Maybe as a followup I'll expose `-j` and when it's `1` it'll use `map` instead of a multiprocessing pool)

```python
$ PATH=/tmp/git:$PATH git-code-debt-generate bagit-java database.db
Cloning into '/tmp/tmpE4e3mYtemp-repo'...
warning: templates not found /home/asottile/share/git-core/templates
done.
Traceback (most recent call last):
  File "/home/asottile/workspace/git-code-debt/venv-git_code_debt/bin/git-code-debt-generate", line 11, in <module>
    load_entry_point('git-code-debt', 'console_scripts', 'git-code-debt-generate')()
  File "/home/asottile/workspace/git-code-debt/git_code_debt/generate.py", line 180, in main
    args.skip_default_metrics,
  File "/home/asottile/workspace/git-code-debt/git_code_debt/generate.py", line 97, in load_data
    commits, map(_get_metrics_inner, mp_args),
  File "/home/asottile/workspace/git-code-debt/git_code_debt/generate.py", line 53, in _get_metrics_inner
    return get_metrics(diff, metric_parsers)
  File "/home/asottile/workspace/git-code-debt/git_code_debt/generate.py", line 38, in get_metrics
    file_diff_stats = get_file_diff_stats_from_output(diff)
  File "/home/asottile/workspace/git-code-debt/git_code_debt/file_diff_stat.py", line 151, in get_file_diff_stats_from_output
    return [_to_file_diff_stat(file_diff) for file_diff in files[1:]]
  File "/home/asottile/workspace/git-code-debt/git_code_debt/file_diff_stat.py", line 107, in _to_file_diff_stat
    assert mode is not None
AssertionError
```

**Comment by asottile** [comment]
Here's the diff it is choking on:

```
(Pdb) print(file_diff[:-1])
 a/trunk/inventory-tools/bin/inventory b/trunk/transfer/bin/inventory
similarity index 100%
rename from trunk/inventory-tools/bin/inventory
rename to trunk/transfer/bin/inventory
```

git2 drastically improved move detection, let's see if I can disable that (might actually speed up generation as well!).   Seems I can disable this by passing `--no-renames` to `git diff`.  This option also exists in 1.9 so I think it's safe to add.  Now to write a regression test and fix it!

**Comment by asottile** [comment]
#64 should fix this.  I've verified that the repository in question generates using the latest version of git with that change

**Comment by asottile** [comment]
This fix should be available in git-code-debt==0.7.2

Thanks for the bug report!

**Comment by dannyroberts** [comment]
Thanks!
